### PR TITLE
Fix a few bugs related to object generation detection

### PR DESCRIPTION
**Pull request by cshung**
@Maoni0 reported that generational aware analysis in PerfView is not working for full core dumps. After debugging for a while, I have found a few bugs related to generation detection. This PR fixes some of them, the remaining ones are in ClrMD and are fixed [there](https://github.com/microsoft/clrmd/pull/1035).

**Note** - I changed the serialization format by adding information about POH, which I need help with backward compatibility. The CI failures are related to this.

@tommcdon - This might need to be replicated for `dotnet-gcdump`

### CODE REVIEW

File: src/HeapDumpCommon/DotNetHeapInfo.cs [hunk 1]
```diff
@@ -136,6 +142,16 @@ void IFastSerializable.FromStream(Deserializer deserializer)
...
+        // 
+        // TODO: Backward compatibility - changing serialization format should require bumping version and code paths to deal with 
+        // missing values. How?
+        //
+        // In case this value was not saved, there could be two cases:
+        // Case 1: We are running on a runtime without POH, in that case, initializing that to Start works.
+        // Case 2: We are running on a runtime with POH, in that case, POH segments were saved as if
+        //         they were Gen2, there is no way to differentiate, so just pretend they are Gen2.
+        //
+        Gen4End = (Address)deserializer.ReadInt64();
```
Comment (brianrob) [inline src/HeapDumpCommon/DotNetHeapInfo.cs:154]: You won't be able to distinguish between whether the runtime has a POH or not at deserialization time.  If you need to know this at deserialization time, you'll need to add this information and serialize it.  Based on your comment, I think you're fine to not worry about it.

I recommend that you add an implementation of IFastSerializableVersion to GCHeapDumpSegment.  You should set the following values:

Version=1
MinimumVersionCanRead=0
MinimumReaderVersion=1

These numbers mean that the current code can read old versions without Gen4End serialized as well as the current version with Gen4End serialized.  You'll also need to add a check prior to deserializing Gen4End in FromStream to make sure that you only deserialize Gen4End if deserializer.VersionBeingRead >= 1.

Take a look at [https://github.com/microsoft/perfview/blob/main/src/FastSerialization/FastSerialization.cs#L2326](https://github.com/microsoft/perfview/blob/main/src/FastSerialization/FastSerialization.cs#L2326) for more details.

**Review by brianrob — APPROVED** [review]
LGTM.  @cshung, are you ready to merge this?

**Comment by cshung** [comment]
> LGTM. @cshung, are you ready to merge this?

Yes, this is ready to be merged. I don't have access to merge it.

**Comment by brianrob** [comment]
> > LGTM. @cshung, are you ready to merge this?
> 
> Yes, this is ready to be merged. I don't have access to merge it.

Yup, no worries.  Just wanted to make sure that I didn't merge too soon.

**Pull request merged**